Add WorkerLatencyProfile enhancement (#977)
openshift/enhancements: enhancements/worker-latency-profile/worker-latency-profile.md

## Design Details

### API Extensions

The [existing proposed node object](https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/mco-cgroupsv2-support.md#api-extensions) needs to be updated to include `NodeStatus` to track the progress of `WorkerLatencyProfile` rollout.

```go
type Node struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// +required
	Spec NodeSpec `json:"spec"`

  // status holds observed values.
	// +optional
	Status NodeStatus `json:"status"`
}

type NodeSpec struct {
  WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`

  // an eventual additional option might be crun in the future. This explains
  //   why a new struct may be necessary
  // CgroupMode CgroupMode `json:"cgroupMode,omitempty"`
  //
  // CrunEnabled bool ...
}

type NodeStatus struct {
  // WorkerLatencyProfileStatus provides the current status of WorkerLatencyProfile
  WorkerLatencyProfileStatus WorkerLatencyProfileStatus `json:"workerLatencyProfileStatus,omitempty"`
}

type WorkerLatencyProfileType string

const (
    // Medium Update and Average Reaction

>     MediumUpdateAndAverageReaction WorkerLatencyProfileType = "MediumUpdateAndAverageReaction"

Reviewer: Nit: drop the and from the name and make it just `MediumUpdateAverageReaction` and `LowUpdateSlowReaction` and provide a better description above the field, ie: `Medium Kubelet Update Frequency (heart-beat) and Average Reaction Time to unresponsive Node` same the other one.

## Summary

To make sure the Openshift cluster keeps running optimally where network latency between the control plane and the worker nodes may not always be at its best (e.g. Edge cases), we can tweak the frequency of the status updates done by the `Kubelet` and the corresponding reaction times of the `Kube Controller Manager` and `Kube API Server`.

## Motivation

To accommodate the higher network latency (such as Edge cases) cluster may experience, we need to adjust how frequently the `Kubelet` updates its status and how components such as `Kube Controller Manager` and `Kube API Server`react to those status updates.

The main motivation of this enhancement is to allow setting relevant arguments for these critical components in a more controlled manner instead of letting the users directly modify them manually which will make the cluster unsupported. This will also eliminate any room for manual errors that could lead to unscheduled downtime.

### Goals

* Enable Openshift cluster to fine tune the reliability in medium to high latency scenarios by specifying simple `WorkerLatencyProfile` at day-0 as well as day-2.

### Non-Goals

> * Modify the existing `Kubelet`, `Kube Controller Manager` or `Kube API Server` code in any way.

Reviewer: ```suggestion
* Modify the existing `Kubelet`, `Kube Controller Manager` or `Kube API Server` configuration in any way.
```

Author: I actually meant `code` only. IMHO, we are indeed changing the configuration of the `Kubelet`, `Kube Controller Manager` or `Kube API Server` by changing the values of the respective arguments (e.g. `--node-monitor-grace-period` arg of `Kube Controller Manager`)

* Allow any kind of different latency scenarios between masters

### User Stories

* User wants to fine tune the cluster reliability for their node latency scenario while making sure they are protected from setting parameters that could potentially break the cluster.

## Proposal

> * The option to set `WorkerLatencyProfile` will have to reside in a centralized location. The [OpenShift Infrastructure config object](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L28) contains information describing how a cluster functions including cloud config and platform specification for each cloud. Setting the  `WorkerLatencyProfile` is an infrastructure setting.

Reviewer: Here you're describing and pointing at `InfrastructureSpec` but in the API Extensions section you're adding `WorkerLatencyProfile` in the `Node` resource, which one you're planning to go with?

Author: Thanks for pointing that out, that paragraph remained from a previous edit that should have been removed. I cleaned it up. Thanks again.

> * [MCO](https://github.com/openshift/machine-config-operator) set the appropriate value of the `Kubelet` flag `--node-status-update-frequency`

Reviewer: ```suggestion
* [machine-config-operator (MCO)](https://github.com/openshift/machine-config-operator) sets the appropriate value of the `Kubelet` flag `--node-status-update-frequency`
```
try not to use acronyms you haven't introduced before. This applies to the next two as well.

Author: Updated, thanks.

* [KCMO](https://github.com/openshift/cluster-kube-controller-manager-operator) set the appropriate value of the `Kube Controller Manager` flag `--node-monitor-grace-period`

> * [CKAO](https://github.com/openshift/cluster-kube-apiserver-operator)set the appropriate value of the `Kube API Server` flags `--default-not-ready-toleration-seconds` and `--default-unreachable-toleration-seconds`

Reviewer: `KASO` is the preferred abbreviation.

Author: Updated, thanks.

## Design Details

### API Extensions

```go
type WorkerLatencyProfileType string

const (
    // Medium Update and Average Reaction
    MediumUpdateAndAverageReaction WorkerLatencyProfileType = "MediumUpdateAndAverageReaction"

    // Low Update and Slow Reaction
    LowUpdateAndSlowReaction WorkerLatencyProfileType = "LowUpdateAndSlowReaction"

    // Default values of relavent Kubelet, Kube Controller Manager and Kube API Server
    Default WorkerLatencyProfileType = "Default"
)

> type Node struct {

Reviewer: This code has too many information, focus only on new additions, so it's more readable.

Author: I removed the declaration of `ConditionStatus` and `ObjectReference` because they are generic enough. But I am afraid  dropping anything more than would lead to loss of information that could be helpful during implementation.

)
type ConditionStatus string

// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
// can't decide if a resource is in the condition or not.
const (
	ConditionTrue    ConditionStatus = "True"
	ConditionFalse   ConditionStatus = "False"
	ConditionUnknown ConditionStatus = "Unknown"
)

type WorkerLatencyStatusConditionOwner string

const (
  // Machine Config Operator will update condition status by setting this as owner
  MachineConfigOperator WorkerLatencyStatusConditionOwner = "MachineConfigOperator"

  // Kube Controller Manager Operator will update condition status  by setting this as owner
  KubeControllerManagerOperator WorkerLatencyStatusConditionOwner = "KubeControllerManagerOperator"

  // Kube API Server Operator will update condition status by setting this as owner
  KubeAPIServerOperator WorkerLatencyStatusConditionOwner = "KubeAPIServerOperator"
)

type WorkerLatencyStatusCondition struct {
	// Owner specifies the operator that is updating this condition
	// +kubebuilder:validation:Required
	// +required

> 	Owner WorkerLatencyStatusConditionOwner string `json:"owner"`

Reviewer: IIRC last time we spoke, @deads2k proposed that each operator will report that status inside its own Status, did that change?

Author: IIRC, @deads2k mentioned each operator reports the status inside it's own `Status` even today. But I thought that was general statement and wasn't binding to this proposal. 

When the user rolls out `WorkerLatencyProfile`, they need to know the status of that roll out. Any success or failure of an individual component reported in a centralized place will be extremely useful for the user. Such information is also useful for performing a rollback (e.g. if kubelet args fail to get updated, controller manager and API server args can be rolled back by their respective operators by observing this centralized status)

	// +kubebuilder:validation:Required
	// +required
	LastTransitionTime metav1.Time `json:"lastTransitionTime"`

	// reason is the CamelCase reason for the condition's current status.
	// +optional
	Reason string `json:"reason,omitempty"`

	// message provides additional information about the current condition.
	// This is only to be consumed by humans.  It may contain Line Feed
	// characters (U+000A), which should be rendered as new lines.
	// +optional
	Message string `json:"message,omitempty"`
}

// ObjectReference contains enough information to let you inspect or modify the referred object.
type ObjectReference struct {
	// group of the referent.
	// +kubebuilder:validation:Required
	// +required
	Group string `json:"group"`
	// resource of the referent.
	// +kubebuilder:validation:Required
	// +required
	Resource string `json:"resource"`
	// namespace of the referent.
	// +optional
	Namespace string `json:"namespace,omitempty"`
	// name of the referent.
	// +kubebuilder:validation:Required
	// +required
	Name string `json:"name"`
}
```

### Operational Aspects of API Extensions

#### Default Update And Default Reaction

> Openshift ships with following default configuration which works quite well in most cases.

Reviewer: ```suggestion
OpenShift ships with the following default configuration which works in most cases.
```

Author: Updated, thanks.

Pods with `NoExecute` taint will get executed as per [tolerationSeconds](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions), but pods without such taint will get evicted in 300 seconds (`--default-not-ready-toleration-seconds` and `--default-unreachable-toleration-seconds` settings of the `Kube API Server`)

| Component               | Flag Name                                | Flag Value  |
| -----------             | -----------                              | ----------- |
| Kubelet                 | --node-status-update-frequency           |    10s      |
| Kube Controller Manager | --node-monitor-grace-period              |    40s      |
| Kube API Server         | --default-not-ready-toleration-seconds   |    300s     |
| Kube API Server         | --default-unreachable-toleration-seconds |    300s     |

#### Medium Update And Average Reaction

> While the default configuration works well in most cases, sometimes the worker nodes may find themselves in a network with slightly higher than usual latency.

Reviewer: works ~~well~~ in most cases...

Author: Updated, thanks.

`Kube Controller Manager` will wait for 5 minutes to consider unhealthy status for the node and since `--default-not-ready-toleration-seconds` and `--default-unreachable-toleration-seconds` of the `Kube API Server` are set to 60 seconds the total time will be 6 minutes before eviction process starts.

| Component               | Flag Name                                | Flag Value  |
| -----------             | -----------                              | ----------- |
| Kubelet                 | --node-status-update-frequency           |    1m       |
| Kube Controller Manager | --node-monitor-grace-period              |    5m       |
| Kube API Server         | --default-not-ready-toleration-seconds   |    60s      |
| Kube API Server         | --default-unreachable-toleration-seconds |    60s      |

#### Failure Modes

> 1. In case of failure `WorkerLatencyProfileStatus` should point towards failed component(s)

Reviewer: See my other comment about reporting status, in the API section.

Author: Thanks, I replied there https://github.com/openshift/enhancements/pull/977#discussion_r782692490

### Cluster Stability Analysis

Let's try to understand the effect of applying various worker latency profiles on the overall stability of the cluster.

> Kubelet will try to make [5 post attempts](https://github.com/kubernetes/kubernetes/blob/release-1.23/pkg/kubelet/kubelet.go#L129) to update it's status. `--node-status-update-frequency` determines the frequency of attemps to update status.

Reviewer: In the default installation, kubelet will...

Author: When it comes to number of  attempts, the value is not configurable in the kubelet. It's [hard coded](https://github.com/kubernetes/kubernetes/blob/release-1.23/pkg/kubelet/kubelet.go#L129) to 5 attempts in the source code. So I am hesitant to state `In the default installation`, because it's unconfigurable for default or non-default configuration of the Kubelet.

| 10s                 | 40s           |    60s       |    No major impact on cluster health
| 10s                 | 5m            |    300s       |  No major impact on cluster health
| 10s                 | 5m           |     60s       | No major impact on cluster health
| 10s                 | 40s           |    300s       |   Default Profile (Final state)

For day-2 operation this transtion will result in **LowUpdateAndSlowReaction -> MediumUpdateAndAverageReaction -> Default** to to minimize the impact on the stability of the cluster.

#### Support Procedures

### Test Plan

Testing should be thoroughly done at all levels, including unit, end-to-end, and integration.

### Graduation Criteria

#### Dev Preview -> Tech Preview

`WorkerLatencyProfile` will be dev preview on its initial release. Internal and customer
usage will be critical to gather information on bugs and enhancements to the
underlying subsystem.

Graduation requirements to Tech Preview are:

* No regressions after applying the `WorkerLatencyProfile` in bringing up `Kubelet`, `Kube Controller Manager` and `Kube API Server`

> * No major imapct on the cluster stability while applying the `WorkerLatencyProfile`, especially nodes should not incorrectly get NotReady/Unreachable taint for prolonged period.

Reviewer: except for the described disruptions as described in the previous section during rollout

Author: Updated, thanks.
